diff --git a/src/Plugin/search_api/backend/SearchApiSolrBackend.php b/src/Plugin/search_api/backend/SearchApiSolrBackend.php
index 8cba690..c3e54b9 100644
--- a/src/Plugin/search_api/backend/SearchApiSolrBackend.php
+++ b/src/Plugin/search_api/backend/SearchApiSolrBackend.php
@@ -2525,7 +2525,7 @@ class SearchApiSolrBackend extends BackendPluginBase implements SolrBackendInter
 
     // Execute the query.
     $result = $connector->extract($query);
-    return $connector->getContentFromExtractResult($result, $filepath);
+    return $connector->getContentFromExtractResult($result, basename($filepath));
   }
 
 }
